bucket notifications - events for scale

alphaprinz opened the pull request:
### Explain the changes
1. Adds events to scenarios the scale wishes to monitor.
2. New event for a failed notification.
3. New event for low space in monitoring logs FS.

Free space below threshold sample:
Jan-26 10:56:06.747 [nsfs/573237] [EVENT]{"timestamp":"2025-01-26T18:56:06.746Z","host":"li-56a4e0cc-353c-11b2-a85c-9bc45442ed4b.ibm.com","event":{"code":"noobaa_notification_low_space","message":"Pending notification log dir low on space","description":"Low space","entity_type":"NODE","event_type":"WARN","scope":"NODE","severity":"WARN","state":"HEALTHY","arguments":{"fs_stat":{"type":1481003842,"bsize":4096,"blocks":483087906,"bfree":411722363,"bavail":411722363,"files":193329536,"ffree":185506484}},"pid":573237}}

Failed notification sample:
Jan-26 11:05:28.657 [/576243] [EVENT]{"timestamp":"2025-01-26T19:05:28.657Z","host":"li-56a4e0cc-353c-11b2-a85c-9bc45442ed4b.ibm.com","event":{"code":"noobaa_notification_failed","message":"Failed to send notification., value: first","description":"Notification failed., error: Error: No such file or directory","entity_type":"NODE","event_type":"WARN","scope":"NODE","severity":"WARN","state":"HEALTHY","arguments":{"code":"ENOENT","context":"Readfile _path=/etc/noobaa.conf.d/connections/connect_http.json "},"pid":576243}}

### Issues: Fixed #xxx / Gap #xxx
1. -

### Testing Instructions:
1. Achieve or simulate low FS space. Alternatively, set a high threshold for event (eg NOTIFICATION_SPACE_CHECK_THRESHOLD = 0.99)
2. Set NOTIFICATION_REQ_PER_SPACE_CHECK to a low value, but more than 0 (eg 1).
3. Setup notification, send a notification
5. Event should appear in logs.


- [ ] Doc added/updated
- [ ] Tests added

alphaprinz committed:
bucket notifications - events for scale

Signed-off-by: Amit Prinz Setter <alphaprinz@gmail.com>

shirady commented on docs/NooBaaNonContainerized/ConfigFileCustomizations.md:
```diff
@@ -478,6 +478,34 @@ Warning: After setting this configuration, NooBaa will skip schema validations a
+    1. Open /path/to/config_dir/config.json file.
+    2. Set the config key -
+    Example:
+    "NOTIFICATION_REQ_PER_SPACE_CHECK": 1000
```
What does this 1000 mean? Could you write it explicitly in the example description?

shirady commented on docs/NooBaaNonContainerized/ConfigFileCustomizations.md:
```diff
@@ -478,6 +478,34 @@ Warning: After setting this configuration, NooBaa will skip schema validations a
+    2. Set the config key -
+    Example:
+    "NOTIFICATION_REQ_PER_SPACE_CHECK": 1000
+    3. systemctl restart noobaa
```
Are you sure you need to restart the service for this?

alphaprinz replied:
How else config.js will be reloaded? Is there a way to reload it while endpoint is running?

shirady replied:
I'm just asking because there are cases in this configs where it was not mentioned.

shirady commented on src/util/notifications_util.js:
```diff
@@ -494,10 +499,31 @@ function get_notification_logger(locking, namespace, poll_interval) {
+    req.notification_logger.writes_counter += 1;
+
+    if (req.notification_logger.writes_counter > config.NOTIFICATION_REQ_PER_SPACE_CHECK) {
+        req.notification_logger.writes_counter = 0;
+        const fs_stat = fs.statfsSync(config.NOTIFICATION_LOG_DIR);
+        if (fs_stat.bavail / fs_stat.blocks < config.NOTIFICATION_SPACE_CHECK_THRESHOLD) {
```
Why at start do you init the `writes_counter` to 0?
Could you explain in a comment the logic for the calc?

alphaprinz replied:
writes_counter counts how many writes has been performed since last check.
it starts at zero, and then zeroed on each check.

shirady commented on src/util/notifications_util.js:
```diff
@@ -145,6 +147,9 @@ class Notificator {
             } catch (err) {
                 dbg.error("Failed to notify. err = ", err, ", str =", str);
+                if (this.nc_config_fs) {
+                    new NoobaaEvent(NoobaaEvent.NOTIFICATION_FAILED).create_event(null, {name: notif?.meta?.name}, err);
```
I saw in other implementations that they used `undefined` instead of `null`.
```suggestion
                    new NoobaaEvent(NoobaaEvent.NOTIFICATION_FAILED).create_event(undefined, {name: notif?.meta?.name}, err);
```

In case the name is `undefined` (`notif?.meta?.name`) would you want to pass additional arguments?

alphaprinz replied:
I've switched to `underfined`.

I don't think more details are needed. `name` is not available in an unlikely case.

shirady commented on src/util/notifications_util.js:
```diff
@@ -494,10 +499,31 @@ function get_notification_logger(locking, namespace, poll_interval) {
+        req.notification_logger.writes_counter = 0;
+        const fs_stat = fs.statfsSync(config.NOTIFICATION_LOG_DIR);
+        if (fs_stat.bavail / fs_stat.blocks < config.NOTIFICATION_SPACE_CHECK_THRESHOLD) {
+            new NoobaaEvent(NoobaaEvent.NOTIFICATION_LOW_SPACE).create_event(null, {fs_stat});
```
I saw in other implementations that they used `undefined` instead of `null`.
```suggestion
            new NoobaaEvent(NoobaaEvent.NOTIFICATION_LOW_SPACE).create_event(undefined, {fs_stat});
```
Do you need all the stat information?

alphaprinz replied:
It's just for logs (ie, not needed), might as well include it, no?

shirady replied:
Whatever you think.
Could you please attach to the PR description the printings of the new events (so we can see a full example).

shirady commented on docs/NooBaaNonContainerized/ConfigFileCustomizations.md:
```diff
@@ -478,6 +478,34 @@ Warning: After setting this configuration, NooBaa will skip schema validations a
 
+### 33. Notification space monitor frequency flag -
+* <u>Key</u>: `NOTIFICATION_REQ_PER_SPACE_CHECK`
+* <u>Type</u>: Number
```
Any limitation on this number? (only positive, not higher than...) you can write in the description.

shirady commented on docs/NooBaaNonContainerized/ConfigFileCustomizations.md:
```diff
@@ -478,6 +478,34 @@ Warning: After setting this configuration, NooBaa will skip schema validations a
+### 34. Notification space monitor threshold flag -
+* <u>Key</u>: `NOTIFICATION_SPACE_CHECK_THRESHOLD`
+* <u>Type</u>: Number
+* <u>Default</u>: 0.1
```
Possible values that can be configured?

nadavMiz commented on src/util/notifications_util.js:
```diff
@@ -115,9 +116,10 @@ class Notificator {
+        let notif;
         await file.collect_and_process(async str => {
             try {
-                const notif = JSON.parse(str);
```
why did you change this? where do you write to this value?

alphaprinz replied:
The `notif` declaration is outside the `try` block so that is could be used for to get the name for the event in the `catch` block.

nadavMiz commented on src/manage_nsfs/manage_nsfs_events_utils.js:
```diff
@@ -384,4 +384,30 @@ NoobaaEvent.CONFIG_DIR_UPGRADE_FAILED = Object.freeze({
+    state: 'HEALTHY',
+});
+
+NoobaaEvent.NOTIFICATION_FAILED = Object.freeze({
```
would this be relevant when other notifications will fail as well?

alphaprinz replied:
I'm not sure I understand.
You mean notifications other than bucket event notifications?
If so, maybe? It's possible, other places would use it for different purposes?
I can change the name to BUCKET_NOTIFICATION_FAILED if that would help.

nadavMiz replied:
you can change it to BUCKET_NOTIFICATION_FAILED if you think it needs to be used only in this context. but maybe we should send this notification in other places. we can also do it in the future on a different PR

aspandey commented on src/util/notifications_util.js:
```diff
@@ -494,10 +499,31 @@ function get_notification_logger(locking, namespace, poll_interval) {
+        req.notification_logger.writes_counter = 0;
+        const fs_stat = fs.statfsSync(config.NOTIFICATION_LOG_DIR);
+        if (fs_stat.bavail / fs_stat.blocks < config.NOTIFICATION_SPACE_CHECK_THRESHOLD) {
+            new NoobaaEvent(NoobaaEvent.NOTIFICATION_LOW_SPACE).create_event(null, {fs_stat});
```
Okay, this we need to think...

Once we reach the space limit where we have to raise an event, means we are in danger of running out of space?
If so, this write onward, we should raise this event after every write and not after "NOTIFICATION_REQ_PER_SPACE_CHECK"

so I think we sould reset req.notification_logger.writes_counter = 0; 
if we are okay with the space and 

if (fs_stat.bavail / fs_stat.blocks < config.NOTIFICATION_SPACE_CHECK_THRESHOLD)

is not true.

alphaprinz replied:
I'm not sure, sounds excessive logging for a situation that is not necessarily wrong or avoidable, where functionality is not necessarily degrade.
I would guess that if users ignore the warn event writing a lot more warn event will probably won't help.

shirady commented on docs/NooBaaNonContainerized/ConfigFileCustomizations.md:
```diff
@@ -478,6 +478,34 @@ Warning: After setting this configuration, NooBaa will skip schema validations a
 
+### 33. Notification space monitor frequency flag -
+* <u>Key</u>: `NOTIFICATION_REQ_PER_SPACE_CHECK`
+* <u>Type</u>: Positive integer
```
I think it was fine "Number" (I looked at existing example like `GPFS_DOWN_DELAY` and `NSFS_DIR_CACHE_MAX_TOTAL_SIZE`), and mention that it is a positive integer.

shirady commented on docs/NooBaaNonContainerized/ConfigFileCustomizations.md:
```diff
@@ -478,6 +478,34 @@ Warning: After setting this configuration, NooBaa will skip schema validations a
+* <u>Key</u>: `NOTIFICATION_SPACE_CHECK_THRESHOLD`
+* <u>Type</u>: Number
+* <u>Default</u>: 0.1
+* <u>Description</u>: Fraction (more than zero, less than 1) of availble block, below which event is logged.
```
Does it include 0 and 1? (from what I understand here no...).

alphaprinz replied:
No, range is open.

shirady commented on docs/NooBaaNonContainerized/ConfigFileCustomizations.md:
Since you add configuration - I think you should update the NSFS configuration schema and we also have tests for this schema changes, see:
`nsfs_config_schema.js` file
https://github.com/noobaa/noobaa-core/blob/97f8b75e6dd786ae4242bdc0d43c26ebe0d03dd3/src/server/system_services/schemas/nsfs_config_schema.js#L4-L11
`test_nc_nsfs_config_schema_validation.test.js` file, for example:
https://github.com/noobaa/noobaa-core/blob/97f8b75e6dd786ae4242bdc0d43c26ebe0d03dd3/src/test/unit_tests/jest_tests/test_nc_nsfs_config_schema_validation.test.js#L52-L60

shirady commented:
@alphaprinz looking at the examples in the PR description:
1. In the first example I was expecting to see "Free space below threshold sample" but in the message I see "Noobaa started", so we probably need to paste another event here.
2. In the second event "Failed notification" in the message I see "description":"Notification failed., error: Error: No such file or directory", do you think it is enough? there is no information about the file name.

alphaprinz commented:
@shirady 
1. I have fixed the sample.
2. I changed the notification failure event to include the err.toString() with more details (sample was also updated).

shirady commented on src/util/notifications_util.js:
```diff
@@ -514,10 +519,36 @@ function get_notification_logger(locking, namespace, poll_interval) {
+    if (req.notification_logger.writes_counter > config.NOTIFICATION_REQ_PER_SPACE_CHECK) {
+        //yes. remember we've just ran  the check by zero-ing the counter.
+        req.notification_logger.writes_counter = 0;
+        const fs_stat = fs.statfsSync(config.NOTIFICATION_LOG_DIR);
```
I know that in NSFS we use: `await nb_native().fs.stat` to get the stat.
Why did you choose to use stat from fs and in its sync way?
I think it's worth adding a comment about it.

shirady commented on src/util/notifications_util.js:
```diff
@@ -514,10 +519,36 @@ function get_notification_logger(locking, namespace, poll_interval) {
+        //yes. remember we've just ran  the check by zero-ing the counter.
+        req.notification_logger.writes_counter = 0;
+        const fs_stat = fs.statfsSync(config.NOTIFICATION_LOG_DIR);
+        //is the ratio of available blocks less than the configures threshold?
+        if (fs_stat.bavail / fs_stat.blocks < config.NOTIFICATION_SPACE_CHECK_THRESHOLD) {
```
Why did you choose to use `bavail` (Free blocks available to unprivileged users) and not `bfree` (Free blocks in file system) ([nodejs documentation](https://nodejs.org/api/fs.html#statfsbavail))?
I think it's worth adding a comment about it (you can expand the existing comment).

shirady reviewed (approved):
LGTM
